File: Java/BucketSort.java
import java.io.FileReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.io.FileOutputStream;

// Average complexity: O(n+k)
// Worst case complexity: O(n^2)

public class BucketSort {
    public static int[] readDataset(String filePath, int maxSize) {
        Scanner datasetFile;
        try {
            datasetFile = new Scanner(new File(filePath));
        } catch (FileNotFoundException e) {
            System.err.println("Failed to open dataset file");
            return null;
        }

        int[] numbers = new int[maxSize];
        int index = 0;
        while (datasetFile.hasNextInt() && index < maxSize) {
            numbers[index++] = datasetFile.nextInt();
        }
        datasetFile.close();

        return numbers;
    }

    public static void main(String args[]) {
        // Default values
        String filePath = "../Dataset/1000k_parc_ordenado.txt";
        String resultPath = "../Results/BucketSort.txt";
        int maxSize = 1000000;

        // Check if the user provided the input file
        if (args.length >= 1) {
            filePath = args[0];
        }

        if (args.length >= 2) {
            resultPath = args[1];
        }

        if (args.length >= 3) {
            maxSize = Integer.parseInt(args[2]);
        }

        int[] numbers = readDataset(filePath, maxSize);
        if (numbers == null) {
            return;
        }

        long startTime = System.currentTimeMillis();
        bucketSort(numbers);
        long endTime = System.currentTimeMillis();

        long elapsedTime = endTime - startTime;

        System.out.println("Elapsed time: " + elapsedTime + " miliseconds");

        writeResult(elapsedTime, filePath, resultPath);
    }

    public static void bucketSort(int[] numbers) {
        int max = Arrays.stream(numbers).max().getAsInt();
        int min = Arrays.stream(numbers).min().getAsInt();
        int bucketSize = 5;
        int bucketCount = (max - min) / bucketSize + 1;
        List<List<Integer>> buckets = new ArrayList<>(bucketCount);
        for (int i = 0; i < bucketCount; i++) {
            buckets.add(new ArrayList<>());
        }

        for (int number : numbers) {
            int bucketIndex = (number - min) / bucketSize;
            buckets.get(bucketIndex).add(number);
        }

        int index = 0;
        for (List<Integer> bucket : buckets) {
            int[] bucketArray = new int[bucket.size()];
            for (int i = 0; i < bucket.size(); i++) {
                bucketArray[i] = bucket.get(i);
            }
            insertionSort(bucketArray);
            for (int number : bucketArray) {
                numbers[index++] = number;
            }
        }
    }

    public static void insertionSort(int[] numbers) {
        for (int currentIndex = 1; currentIndex < numbers.length; currentIndex++) {
            int currentNumber = numbers[currentIndex];
            int previousIndex = currentIndex - 1;
            while (previousIndex >= 0 && numbers[previousIndex] > currentNumber) {
                numbers[previousIndex + 1] = numbers[previousIndex];
                previousIndex = previousIndex - 1;
            }
            numbers[previousIndex + 1] = currentNumber;
        }
    }

    public static void writeResult(long executionTime, String fileName, String resultPath) {
        File resultsFile = new File(resultPath);
        PrintWriter writer;
        try {
            writer = new PrintWriter(new FileOutputStream(resultsFile, true)); // true for append mode
        } catch (FileNotFoundException e) {
            System.err.println("Error opening the results file");
            return;
        }

        String name = new File(fileName).getName();
        writer.printf("Java - BucketSort - File: %s\n", name);
        writer.printf("Execution time: %.2f ms\n", (double)executionTime / 1_000_000); // convert nanoseconds to milliseconds
        writer.close();
    }
}

File: C++/BucketSort.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <chrono>

// Function to sort a vector using the Quick Sort algorithm
void quickSort(std::vector<int>& inputVector, int start, int end) {
    if (start < end) {
        int pivot = inputVector[end];
        int smallerElementIndex = start - 1;

        for (int currentIndex = start; currentIndex < end; currentIndex++) {
            if (inputVector[currentIndex] <= pivot) {
                smallerElementIndex++;
                std::swap(inputVector[smallerElementIndex], inputVector[currentIndex]);
            }
        }
        std::swap(inputVector[smallerElementIndex + 1], inputVector[end]);

        int pivotIndex = smallerElementIndex + 1;
        quickSort(inputVector, start, pivotIndex - 1);
        quickSort(inputVector, pivotIndex + 1, end);
    }
}

// Bucket Sort function
std::vector<int> BucketSort(const std::vector<int>& inputVector) {

    int maxValue = *std::max_element(inputVector.begin(), inputVector.end());
    int minValue = *std::min_element(inputVector.begin(), inputVector.end());

    // Determine the number of buckets
    int bucketCount = (maxValue - minValue) / 10 + 1;

    // Create buckets
    std::vector<std::vector<int>> buckets(bucketCount);

    // Distribute elements into buckets
    for (int i = 0; i < inputVector.size(); i++) {
        int bucketIndex = (inputVector[i] - minValue) / 10;
        buckets[bucketIndex].push_back(inputVector[i]);
    }

    // Sort elements in each bucket using Quick Sort
    for (int i = 0; i < buckets.size(); i++) {
        quickSort(buckets[i], 0, buckets[i].size() - 1);
    }

    // Concatenate sorted buckets into a single vector
    std::vector<int> sortedVector;
    for (int i = 0; i < buckets.size(); i++) {
        for (int j = 0; j < buckets[i].size(); j++) {
            sortedVector.push_back(buckets[i][j]);
        }
    }

    return sortedVector;
}

void writeResult(double executionTime, std::string fileName, std::string resultPath) {
    FILE *resultsFile;
    resultsFile = fopen(resultPath.c_str(), "a");
    if (resultsFile == NULL) {
        perror("Erro ao abrir o resultsFile");
        return;
    }

    std::string name = fileName.substr(fileName.find_last_of('/') + 1);
    //fprintf(resultsFile, "C - BucketSort - File: %s\n", name);
    fprintf(resultsFile, "Execution time: %.6f ms\n", executionTime);
    fclose(resultsFile);
}

int main(int argc, char *argv[]) {
    // Default values
    std::string filePath = "../../Dataset/1000k_parc_ordenado.txt";
    std::string resultPath = "../../Results/BucketSort.txt";

    size_t maxSize = 1000000;

    // Check if the user provided the input file
    if (argc >= 2) {
        filePath = argv[1];
    }

    if (argc >= 3) {
        resultPath = argv[2];
    }

    if (argc >= 4) {
        maxSize = std::stoi(argv[3]);
    }

    std::ifstream datasetFile(filePath.c_str());

    if (!datasetFile) {
        std::cerr << "Failed to open dataset file" << std::endl;
        return 1;
    }

    int number;
    std::vector<int> numbers;
    while (datasetFile >> number) {
        numbers.push_back(number);
        if (numbers.size() == maxSize) {
            break;
        }
    }
    datasetFile.close();

    // Ordenar numbers usando Bucket Sort
    auto startTime = std::chrono::high_resolution_clock::now();

    std::vector<int> result = BucketSort(numbers);

    auto endTime = std::chrono::high_resolution_clock::now();

    auto executionTime = std::chrono::duration<double, std::milli>(endTime - startTime);
    std::cout << "Execution time: " << std::fixed << executionTime.count() << " milliseconds" << std::endl;

    // std::cout << "numbers Ordenado:" << std::endl;
    // for (int i = 0; i < result.size(); i++) {
    //     std::cout << result[i] << " ";
    // }
    // std::cout << std::endl;

    writeResult(executionTime.count(), filePath, resultPath);

    return 0;
}